### DISCUSSION TITLE
Digilocker integration

### DISCUSSION THREAD

**Discussion by coolbung** [Sunbird-RC/community#234] [Ideas & Enhancements]
What would be the approach to build a digilocker (https://www.digilocker.gov.in/) integration that will allow a user to import documents from digilocker, and add them to their profile in the registry. The document/metadata fetched from digilocker will be added as a claim that will be auto-attested. 

Example, consider a registry with the below User schema (representative schema, may not be accurately generalised)
- name
- dob
- education (attest-able attribute)
  - type (school/degree/diploma)
  - course
  - stream
  - year of passing
  - score
  - institute
  - university

The user should be able to connect his digilocker account, and import documents (12th standard certificates, degree/diploma certificates etc) and add it to the education attribute.

**Comment by pramodkvarma** [comment]
Thanks for initiating this. Few comments:
1. Integration to Digilocker or any credential wallet like system should be loosely coupled through configurable adapters. 
2. Integration has to be both ways. One to push credentials issued via Sunbird-RC to credential wallet systems like Digilocker and reverse (as your use case below) to pull credentials and attach to my profile.

Given Apple and Google have launched credential wallet as part of mobile OS, users may choose any of those and hence all such integration must be configurable, through standard integration API, and with pluggable adapters.

**Comment by tejash-jl** [comment]
SunbirdRC provides integration with external systems to access documents/VCs through adaptors/plugins. We have implemented similar functionality with MOSIP to get the user identity credentials.
Flow: 
- User logins to SunbirdRC
- User initiates fetching MOSIP documents
- User needs to authenticate with the MOSIP system to generate the token
- Using the token the user ID credentials will be imported to SUNBIRD-RC and self-attested

**Reply by pramodkvarma** [reply]
Awesome! Thanks @tejash-jl 
What about reverse integration of pushing Sunbird-RC issued credentials to go into Digilocker or other credential wallets?

**Reply by tejash-jl** [reply]
@pramodkvarma right now we have exposed  generic apis that can be used to download credentials, but we don't have a mechanism to push credentials to an external system. 
Previously in DIVOC, we had written an adapter for 3rdparty (DigiLocker and aarogya setu) to pull DIVOC credentials.  The same can be added here, but it'll be PULL based approach.

**Reply by pramodkvarma** [reply]
OK, let us see if we can bring the pull based system into Sunbird-RC abstracted from DIVOC. But, subsequently, we can also enable, "publish to digilocker" as a flow. Thanks @tejash-jl

**Reply by coolbung** [reply]
Great. Can you share docs about these APIs, and a note on how the digilocker integration could be done ? 
I believe even Aadhaar ekyc can be implemented on the same lines ?

**Reply by dileepbapat** [reply]
Push API specification is not available in [Digilocker reference](https://img1.digitallocker.gov.in/assets/img/issuer_api/Digital%20Locker%20Issuer%20API%20Specification%20v1.12.pdf) . Will try to explore
